firedrake/cofunction.py

```diff
@@ -79,6 +79,18 @@ def __del__(self):
         if hasattr(self, "_comm"):
             mpi.decref(self._comm)
 
+    def __call__(self, primal):
+        r"""Compute the duality pairing, this is to evaluate this
+        :class:`.Cofunction` at a :class:`.Function`, equivalent to taking the
+        dot product of the vectors of coefficients of the dual and primal
+        arguments.
+
+        :arg primal: the :class:`.Function` to be paired with this :class:`.Cofunction`.
+
+        :returns: the scalar result of the duality pairing.
+        """
+        return firedrake.assemble(ufl.action(self, primal))
```

Comment:
To be more general, I think this should perform symbolic dual-evaluation interpolation #2297 . Does `ufl.action` account for the inner product in your Riesz representer?

Reply:
Hmm, I think I've gotten confused. Please ignore.